src/app_model/expressions/_expressions.py

```diff
@@ -184,28 +181,55 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
         if type(self).__name__ == "Expr":
             raise RuntimeError("Don't instantiate Expr. Use `Expr.parse`")
         super().__init__(*args, **kwargs)
+        self.eval = self.eval_with_callables  # type: ignore[method-assign]
```

Comment:
more the former?  more like I'm really not sold on this whole pattern.  I put it up as an experiment for you all to try, and if you say "yes please I can't live without this feature", then I'll merge it... but otherwise I'd probably not.  So, if you want it, then i'm looking for feedback (@Czaki, that includes you) as to whether you want it enabled all the time by default, opt-in, or opt-out.   And, if one of the opt-patterns, how would you want to opt in/out since you don't like this pattern

Reply:
Hmm well, it's not so much that I can't live without it, but it definitely improves our code in a couple of places where we don't want to update context keys based on events because they're too frequent. I generally really like the idea of "lazily evaluated" context keys, because it allows us to maintain the pattern of "reading from the context always gives you the correct value" without us manually going in and populating context keys just before we need them, which I think eventually could be very messy. 

In terms of opt-in/opt-out I'd be happy to make it opt-in if you think it's overall too dangerous a pattern to make default. I have no specific preference on how to opt-in though.

Reply:
Hm. I case of napari the problem comes from the right click menu where information about being non-empty layer is not properly invalidated. But the problem with normal callable is that it could be evaluated multiple time, that consumes obsolete calls (this is why taller point `functools.cache`)

Evaluating all functions from context may lead to making obsolete calls. So maybe a better option could be implementing the own mapping. Something like:

```python
class ContextMapping(collections.Mapping):
    def __init__(initial_values: collections.Mapping):
        self._base_store = initial_values
        self._store = {}
    
    def __getitem__(self, key):
        if key in self._store:
            return self._store[key]
        if item not in self._base_store:
            raise KeyError("....")
        value = self._base_store["key"]
        if callable(value):
            value = value()
        self._store[key] = value
        return value
        
    ...
```

Then we get single time evaluation, but it is also transparent for current implementation, making it cleaner. 

And also such a mapper could be used in places where it is needed, and in places where context is static we could provide simple dict. 

what did you think?

Reply:
Yes it looks like something testable on napari side. I got this idea when reviewing this PR. @lucyleeow what did you think?

Reply:
Sorry could you expand? Would this new `ContextMapping` be used instead of app-model `Context` in cases where a the 'context key' is a function?
How would the cache clearing for the function work?

Reply:
In the context of right click menu (layer list) `ContextMapping` may be created each time the right click menu is triggered.

Reply:
Ah I see. Stupid question, but where would you define the 'function' context keys? Also in `aboutToShow`?

Reply:
> In the context of right click menu (layer list) ContextMapping may be created each time the right click menu is triggered.

@Czaki how are you seeing this integrate with the existing app-model contexts? Are you imagining the `ContextMapping` holding the "static" context keys of a given context in its `_base_store`? I guess for me the question of "how do you update context keys you don't want to connect to events" merits answering with a general solution in app-model, but if it's easier to implement on the napari side, we can go that way. I just don't want a clunky fix that we then have to maintain forever, if this is a real use-case

Reply:
I was thinking about use it here: 

https://github.com/napari/napari/blob/f640234ca73f77305f3ffd8c09b77dff36c8aa60/napari/_qt/containers/_layer_delegate.py#L325-L326

Where functions like `is_valid_spatial_in_clipboard` could be part of context, so it could be something like:
```
        ctx = ContextMapping(get_context(layer_list))
        self._context_menu.update_from_context(ctx) 
```

Where `is_valid_spatial_in_clipboard` is already part of the context, but as a function, not value.

Reply:
Talking with @DragaDoncila, if you don't think you want to support this in app-model we'll implement this new context in napari and let you know if we run into any concerns.
Feel free to close this PR @tlambert03 !

Reply:
@Czaki any interest in putting in a PR since it was your idea? Happy to have a go if not.

Reply:
I will try